### ISSUE TITLE
not tracking 

### ISSUE THREAD

**Issue by 3073** [kcg2015/Vehicle-Detection-and-Tracking#1]
the tacker is not tracking the sequence of images in "test_mages" folder.
the position of the bounding boxes do not change. it well track the  "frame003" image

**Comment by kcg2015** [comment]
Could you provide more detail (error messages or images?)? I assume that you ran the main.py. I ran the example again. It seems to work fine. There could be a few occations that the detector could  miss detecting the car. In these cases, the most recently updated bounding box would be used for tracking (e.g.,  Frame 5).  Was this the situation that you encountered?

Here are a few examples from my end:
1)  measurement and update for Frame 4, respectively:
![frame4_measurement](https://user-images.githubusercontent.com/12797392/36643483-05b9d188-1a1a-11e8-958a-d3e8fc423525.png)
![frame4_update](https://user-images.githubusercontent.com/12797392/36643484-05c42304-1a1a-11e8-8821-4df63bda7775.png)
2)  measurement and update for Frame 8, respectively:
![frame8_measurement](https://user-images.githubusercontent.com/12797392/36643512-4cb5bed0-1a1a-11e8-8720-d39adb271fb8.png)
![frame8_update](https://user-images.githubusercontent.com/12797392/36643515-55c30cee-1a1a-11e8-96b2-06e9ebc60c8d.png)

Let me know if this would help.

Best, 
Kyle

**Comment by 3073** [comment]
Ya that is the case. But I run the tracker.py not the main.py.

**Comment by 3073** [comment]
By the way The tensorflow detection api with the SSD.Mobilet architecture
is not accurate than the pure SSD detector in keras.  I tested it in some
images and the score is very less for this API.....Why we use it?   do you
have SSD with a tracker like kalman filter? or do you know any?


thank you

**Comment by 3073** [comment]
[image: Inline image 1]

**Comment by 3073** [comment]
![image](https://user-images.githubusercontent.com/33939573/36656986-8e11c6f2-1a7f-11e8-8a8d-3c8ef706935a.png)

**Comment by kcg2015** [comment]
@3073 , thanks for your comments and  feedback.

1) The code in tracker.py is meant to quickly illustrate the impact noise covariance matrix of an measurement. You can modify the code:
127 images = [plt.imread(file) for file in glob.glob('./test_images/*.jpg')]
128 img=images[3]
to suit your purpose.

2) I agree with you that the detection (localization) quality of SSD-Mobilenet architecture is far from perfect. I use it mainly because the fast inference time (<200 ms on a CPU only laptop).  Of course, you can use other detection architectures/APIs,  subject to the trade-off between accuracy and inference time. In my experience,  there will always be cases that a general-purpose object detection API   could fail.  Thus I plan to optimize SSD-Mobilenet for vehicle detection only with new training images. I will keep you posted on the progress.

3) I am not quite sure about your question "do you have SSD with a tracker like kalman filter". From my experience, SSD only provides the measurement part (by outputing the bounding box in a frame) for Bayesian filtering (could be KF, UKF, or EKF). I don't believe SSD, in its most basic design, has a built-in predict and update processes. I could be wrong.

Hope these could be helpful. 

Best, Kyle

**Comment by 3073** [comment]
Thank you for real. you are a smart developer with satisfying answer for
you customers.

what I am saying is do you know any github code that illustrates. SSD
detection and any kind of tracking algorithm for the detected objects like
in the following link.
   https://github.com/ykamikawa/yolov2

In this link we are using YOLO for detection and kalman filter for tracking
the detected objects. But since YOLO is difficult to understand I need to
adapt SSD. That is my question.

Sorry if I kill your time.   Thank you very very much.


one more thing: if you have time and able to do that for me I will
definitely donate for your time.

**Comment by kcg2015** [comment]
@3073 , I  am only aware of Simple, online, and realtime tracking (SORT) for multi-object tracking in this repo: https://github.com/abewley/sort. Also if you want to dig deep into the concepts/techniques of varios Bayesian filters for tracking, the following repo is a very good source: https://github.com/rlabbe/Kalman-and-Bayesian-Filters-in-Python

Hope these would help.

Best,
Kyle

**Comment by 3073** [comment]
the link below is the perfect example that I need .  but I do not know
what type of detector is used in detector.py
https://github.com/ZidanMusk/experimenting-with-sort

anyways thank you for your time

**Comment by kcg2015** [comment]
I have no idea either. The best way is to contact the author/developer of the repo directly.

**Comment by 3073** [comment]
Thank you


I have tested your gift (
https://github.com/kcg2015/Vehicle-Detection-and-Tracking) and  It is
working for a video file with minor modification to suit my need.
could you help me how  can I use it  Webcam?

**Comment by kcg2015** [comment]
I have not worked on webcam capturing yet. You can check out this YouTube video and the code for more information. https://www.youtube.com/watch?v=MyAOtvwTkT0. Hope this would help.

**Comment by 3073** [comment]
Thank you.
 do you have some time to show me How the Main.py works in your repo?
 I did not understand the codes in:
# Deal with matched detection
# Deal with unmatched detection
# Deal with unmatched tracks

**Comment by kcg2015** [comment]
@3073 , please give me some time to update the Readme file of the repo. I will provide more details about these three code blocks.

**Comment by 3073** [comment]
Thank you
take your time  my  friend

**Comment by 3073** [comment]
Is there a possibility to use this for counting the  tracked persons.  How
that can be possible

thank you

**Comment by kcg2015** [comment]
Yes.  The number of tracked persons equals to matched detections.

**Comment by 3073** [comment]
Thank you
I am really sorry that I make you busy by these too many question.
Normally I am studying this to count the number of people entering a bus.
do you have any idea that a tracked object will not be counted twice?

**Comment by 3073** [comment]
that means how to avoid counting of a person in the remaining frames  if it
is detected and tracked in the first frame?

**Comment by kcg2015** [comment]
@3073, no worry. Because the matching between the detection and the track is one-to-one, i.e, a detection can be matched to only (at most) one track, and a track can have only (at most) one associated detection,  I doubt that a tracked detection will be double-counted. Does it sound reasonable to you? Let me know.

**Comment by 3073** [comment]
by the way excuse me that I do not STAR your repo until now but you
answered me (in fact you teach me)  many questions. I will do that before I
start my work!!!

Ya it sounds good and I will try my best and let you now the result.

thank you for real

**Comment by kcg2015** [comment]
@3073 , I am glad this repo can be helpful to you. Thanks for the star. Again, please bear with me as it takes some time to update README file and some code blocks. This is not my day-job, so progress could be slow.

**Comment by 3073** [comment]
ok. I stick with you

**Comment by 3073** [comment]
Hi
I was testing it to count a person in a video.
 I feed a video that contain a single same person in all the frames.
but it is counting the person many times ( equal to the number of frames
that contain the person)   provide me some ideas

how do I identify the person in the second and third frame is the person
that is detected in the first frame?


Thank you very much

**Comment by kcg2015** [comment]
Hi, @3073, I have never implementd object-counting feature before, so I have don't have any first-hand experience. So my guess is (still) counting the number of matched tracks (or detections) for a given frame.  Also could you post your code (block) on github (public or private at your own choice)? Without going through your code, it is very hard to pinpoint the what happened.

**Comment by 3073** [comment]
Ok thank you

**Comment by 3073** [comment]
Hi how are you.  look below for my test on a video
 what is wrong with the detection box that they double for single person?

**Comment by kcg2015** [comment]
@3073, could you show me an image of multiple boxes on the same object?

**Comment by 3073** [comment]
here is the video

**Comment by kcg2015** [comment]
For some reason,  I can not see your video. See if you can upload it on YouTube and send me a link?

**Comment by 3073** [comment]
Dear my friend 
 do you have some time to show me How the Main.py works in your repo?
 I did not understand the codes in:
# Deal with matched detection
# Deal with unmatched detection
# Deal with unmatched tracks
Lat time you promised to help me understand me all these thing.
off course I understand them nearly. could you be online to show me please?

Thank you

**Comment by kcg2015** [comment]
@3073 , I just update the readme file in the repo. It should be very clear the functionalities of these three code blocks.

**Comment by 3073** [comment]
very nice

**Comment by kcg2015** [comment]
Cool!  I am going to close this issue.

**Comment by 3073** [comment]
Hi there how are you....
I have this issue with the repo
I added dlib tracker with SSD mobile net  and it is not detecting the
person. the detection box is not going in the direction the person is
moving  rather the box is moving down ward.
ca you fix this....? i did not change the code in the detector.py

**Comment by kcg2015** [comment]
@3073 ,  I am afraid that I can not be of much help to you. I am not familiar with dlib tracker at all. Also, from your description, it is very hard to pinpoint what the problem is let alone to fix it. All I can say is that if the performance of SSD Moblenet is not up to your task, consider using a different one in Tensorflow Object Dection API.

**Comment by 3073** [comment]
Thank you very much
I have fixed it......

**Comment by 3073** [comment]
Hi  how are you!!!   have you check any possibility to run this code on raspberry pi?  image classification on raspberry pi using SSD Moblenet .....
